lib/liquid/context.rb

```diff
@@ -161,27 +190,25 @@ def evaluate(object)
 
     # Fetches an object starting at the local scope and then moving up the hierachy
     def find_variable(key, raise_on_not_found: true)
+      variable = UNDEFINED
```

Comment:
Big rewrite of `find_variable` here. The original was... very weird. After verifying that neither `environment` nor `scope` had the variable, it then still looked for them in both -- which would always return `nil`. Now we just fall back to `nil` explicitly.

That is, assuming there's not some behaviour I've missed here. Would appreciate extra :eyes: on this change.

Reply:
I feel like this can be simplified by pulling out these 3 `variable == UNDEFINED` and instead being inside try_find_variable and just return if a match is found or nil if it gets to the end.

Reply:
Problem is that `nil` is a valid value for a defined value :/. That's why I've pulled out `UNDEFINED` separately.

Previously we relied on the `scope` found being `nil` or not to differentiate between a found `nil` and a transient `nil`, but this refactoring doesn't have that.

Reply:
so what I kinda of mean is
```
if index
....
else 
   variable = try_find_variable(raise_on_not_found: raise_on_not_found)
endif
```

```
def try_find_variable( key, raise_on_not_found:) 
   [@environments, @static_environments].each do |environments|
      ...
   end
   nil
end

Reply:
Ah ok, so checking them each there. And since we return `nil` as a default it doesn't matter if we can tell whether it was found or not 👍. I'll try that out.

Reply:
this would avoid needing to use `UNDEFINED` at all

Reply:
That does cost us an allocation technically, though (for the `[@environments, ...]`. Not sure if that's worth worrying about, but this is basically the hottest path in Liquid.

Reply:
if you don't want the allocation I guess you could duplicate the code one after another even though its not pretty

Reply:
Or would ruby support moving this up to the class level

@all_environments = [@environments, @static_environments]

Reply:
:+1: That would also work